src/constbuffer.c

```diff
@@ -394,3 +398,256 @@ IMPLEMENT_MOCKABLE_FUNCTION(, bool, CONSTBUFFER_HANDLE_contain_same, CONSTBUFFER
+                    result = CONSTBUFFER_TO_FIXED_SIZE_BUFFER_RESULT_INSUFFICIENT_BUFFER;
+                }
+                else
+                {
+                    /*Codes_SRS_CONSTBUFFER_02_059: [ CONSTBUFFER_to_fixed_size_buffer shall write at offset 0 of destination the version of serialization (currently 1). ]*/
+                    write_uint8_t(destination + CONSTBUFFER_VERSION_OFFSET, CONSTBUFFER_VERSION);
+
+                    /*Codes_SRS_CONSTBUFFER_02_060: [ CONSTBUFFER_to_fixed_size_buffer shall write at offset 1 of destination the value of source->alias.size in network byte order. ]*/
+                    write_uint32_t(destination + CONSTBUFFER_SIZE_OFFSET, (uint32_t)source->alias.size);
+
+                    /*Codes_SRS_CONSTBUFFER_02_061: [ CONSTBUFFER_to_fixed_size_buffer shall copy all the bytes of source->alias.buffer in destination starting at offset 5. ]*/
+                    (void)memcpy(destination + CONSTBUFFER_CONTENT_OFFSET, source->alias.buffer, source->alias.size);
+
+                    /*Codes_SRS_CONSTBUFFER_02_062: [ CONSTBUFFER_to_fixed_size_buffer shall succeed, write in serialized_size how much it used and return CONSTBUFFER_TO_FIXED_SIZE_BUFFER_RESULT_OK. ]*/
+                    result = CONSTBUFFER_TO_FIXED_SIZE_BUFFER_RESULT_OK;
+                }
+            }
+        }
+    }
+    return result;
+}
+
+CONSTBUFFER_FROM_BUFFER_RESULT CONSTBUFFER_from_buffer(const unsigned char* source, uint32_t size, uint32_t* consumed, CONSTBUFFER_HANDLE* destination)
+{
+    CONSTBUFFER_FROM_BUFFER_RESULT result;
+    if (
+        /*Codes_SRS_CONSTBUFFER_02_063: [ If source is NULL then CONSTBUFFER_from_buffer shall fail and return CONSTBUFFER_FROM_BUFFER_RESULT_INVALID_ARG. ]*/
+        (source == NULL) ||
+        /*Codes_SRS_CONSTBUFFER_02_064: [ If consumed is NULL then CONSTBUFFER_from_buffer shall fail and return CONSTBUFFER_FROM_BUFFER_RESULT_INVALID_ARG. ]*/
+        (consumed == NULL) ||
+        /*Codes_SRS_CONSTBUFFER_02_065: [ If destination is NULL then CONSTBUFFER_from_buffer shall fail and return CONSTBUFFER_FROM_BUFFER_RESULT_INVALID_ARG. ]*/
+        (destination == NULL)
+        )
+    {
+        LogError("invalid arguments const unsigned char* source=%p, uint32_t size=%" PRIu32 ", uint32_t* consumed=%p, CONSTBUFFER_HANDLE* destination=%p", 
+            source, size, consumed, destination);
+        result = CONSTBUFFER_FROM_BUFFER_RESULT_INVALID_ARG;
+    }
+    else
+    {
+        if (size == 0)
+        {
+            /*Codes_SRS_CONSTBUFFER_02_066: [ If size is 0 then CONSTBUFFER_from_buffer shall fail and return CONSTBUFFER_FROM_BUFFER_RESULT_INVALID_ARG. ]*/
+            LogError("cannot deserialize from size=%" PRIu32 "", size);
+            result = CONSTBUFFER_FROM_BUFFER_RESULT_INVALID_ARG;
+        }
+        else
+        {
+            uint8_t version;
+            read_uint8_t(source + CONSTBUFFER_VERSION_OFFSET, &version);
+            if (version != CONSTBUFFER_VERSION)
```

Comment:
>CONSTBUFFER_VERSION [](http://example.com/codeflow?start=27&length=19)

Technically this is the serializer for v1. When v2 would appear I would bet that the v1 serializer will still have to exist (because that is the only way we can do upgrades without interruption of service). So maybe we should consider having constbuffer_serializer_v1 as a unit. And this unit would only do v1 serialization/deserialization. Hardcoded. So when v2 comes along we can slam a v2 next to it.

Of course we can also do the rename when v2 appears :-). Delay until maybe not needed.

Reply:
I will rename this macro to CONSTBUFFER_VERSION_V1. And CONSTBUFFER_VERSION_V2 will come when time comes.